[RDY] man.vim: Simplify man#init to reduce load time

nhooyr opened the pull request:
I removed the SunOS stuff since no one uses SunOS and I've never tested
it on there.

I removed the section_flag init as we can just use -S instead of -s
and -S is used by every implementation as far as I know.

This brings man#init's time from 50-70ms to 20ms for me.

Updates #12318
Related #6766
Related #6815

Perhaps we can make it 0 if we find a way to avoid checking for
the -l flag.

nhooyr committed:
man.vim: Simplify man#init to reduce load time

I removed the SunOS stuff since no one uses SunOS and I've never tested
it on there.

I removed the section_flag init as we can just use -S instead of -s
and -S is used by every implementation as far as I know.

This brings man#init's time from 50-70ms to 15-20ms for me.

Closes #12318
Related #6766
Related #6815

nhooyr commented:
cc @raichoo @justinmk

tjdevries commented:
So I was looking through here and I think it'd be possible to, at the very least, run some of these "system" calls in parallel using `jobstart` and then waiting for the results. Not sure if that would actually improve the speed or not, but it would be something to consider (especially since we're already using `jobstart` to run the commands...)

Additionally, one thing we could do is we could cache the data in `stdpath('cache')` somewhere, and read the results "instantly" from the file if it exists and allow a way to reset those values (perhaps done automatically in some way, but could speed up load time).

Another option would be, instead of removing the checking of any args, we can set up "the happy path" down a line of function calls, and if any of them fail, then we fall back to the other option.

tjdevries commented:
Let me know your thoughts on the above and then I'll look into it more.

nhooyr commented:
> So I was looking through here and I think it'd be possible to, at the very least, run some of these "system" calls in parallel using jobstart and then waiting for the results. Not sure if that would actually improve the speed or not, but it would be something to consider (especially since we're already using jobstart to run the commands...)

That's fair. So I removed the section flag detection because every man implementation actually uses `-S` but some support `-s` as an alias so if we just use `-S`, we won't have any problems. This is actually how the old man.vim plugin worked before I rewrote it, I mistakenly introduced this bug.

See https://github.com/neovim/neovim/pull/4449/commits/42e9606c23eab6542678d2786e7e5819fd704e92

In terms of SunOS, I've never tested it on there and I don't think many people use neovim on SunOS. I think it's very safe to remove.

> Additionally, one thing we could do is we could cache the data in stdpath('cache') somewhere, and read the results "instantly" from the file if it exists and allow a way to reset those values (perhaps done automatically in some way, but could speed up load time).

I like this for the `-l` flag but I think we should do it in a separate PR.

tjdevries commented:
OK. I'm not really sure how to test this. Do you need any more review on it or what can we do?

nhooyr commented:
> OK. I'm not really sure how to test this. Do you need any more review on it or what can we do?

Give me a few days and I'll test this on Debian, macOS and FreeBSD.

Any others you want me to test it on?

tjdevries commented:
Sounds good to me. I didn't check test failures, but should just make sure the travis one is unrelated.

nhooyr commented:
- [x] macOS
- [x] Debian
- [x] FreeBSD

Encountered no issues.

nhooyr commented:
<img width="1118" alt="Screen Shot 2020-07-20 at 1 45 31 AM" src="https://user-images.githubusercontent.com/10180857/87903731-b3129c80-ca2a-11ea-9e37-812589762386.png">

CI failure appears unrelated. That file doesn't use anything from man as far as I can tell.

Also saw this but again seems unrelated.

<img width="1114" alt="Screen Shot 2020-07-20 at 1 46 01 AM" src="https://user-images.githubusercontent.com/10180857/87903769-c6256c80-ca2a-11ea-9248-e3fa977299fa.png">

tjdevries commented:
Re-ran that failing test and it passed, so I think we're good to go! Thanks!
